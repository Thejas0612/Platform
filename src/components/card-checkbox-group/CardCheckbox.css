.cardCheckbox__image{
    max-width: 100%;
    display: block;
    vertical-align: middle;
}