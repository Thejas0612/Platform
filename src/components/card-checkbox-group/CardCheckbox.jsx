import React from "react";
import PropTypes from "prop-types";
import { Box, Checkbox, FormControlLabel, Grid, Tooltip, Typography, useTheme } from "@mui/material";
import styles from "./CardCheckbox.module.css";

export const CardCheckbox = ({
  id,
  name,
  title,
  imageUrl,
  disabled = false,
  disabledTooltip,
  onCompareChange = () => {},
  selectedChecked = false,
  comparedChecked = false,
  onSelectChange = () => {},
}) => {
  const theme = useTheme();
  const disabledColor = theme.palette.action.disabled;

  const HORIZONTAL_CENTER_TOOLTIP = {
    popper: {
      modifiers: [
        {
          name: "offset",
          options: {
            offset: [0, -90],
          },
        },
      ],
    },
  };

  const handleCompareChange = (event) => {
    event.stopPropagation();
    onCompareChange();
  };

  const handleSelectChange = () => {
    if (disabled) {
      return;
    }
    onSelectChange();
  };

  return (
    <Box
      onClick={handleSelectChange}
      sx={{
        cursor: !disabled ? "pointer" : undefined,
        padding: ".5rem",
        border: selectedChecked ? ".125rem solid var(--ddl-color--primary-emerson-green-dark)" : ".0625rem solid var(--ddl-color--primary-grey)",
      }}
    >
      <Grid direction="column" spacing={0.5} alignItems="center" container>
        {/* Title */}
        <Grid item>
          <Typography sx={{ color: disabled ? disabledColor : undefined }}>{title}</Typography>
        </Grid>

        {/* Image and Disabled Tooltip */}
        <Grid item>
          {disabled ? (
            <Tooltip title={disabledTooltip} placement="bottom" slotProps={HORIZONTAL_CENTER_TOOLTIP}>
              <img src={imageUrl} alt={title} className={styles.cardCheckbox__image} />
            </Tooltip>
          ) : (
            <img src={imageUrl} alt={title} className={styles.cardCheckbox__image} />
          )}
        </Grid>

        {/* Checkbox */}
        <Grid item>
          <FormControlLabel
            disabled={disabled}
            onClick={(event) => {
              event.stopPropagation();
            }}
            sx={{ textAlign: "center" }}
            control={<Checkbox onChange={handleCompareChange} checked={comparedChecked} />}
            label="compare"
            name={name}
          />
        </Grid>
      </Grid>
    </Box>
  );
};

CardCheckbox.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  imageUrl: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  disabledTooltip: PropTypes.string,
  onCompareChange: PropTypes.func,
  selectedChecked: PropTypes.bool,
  comparedChecked: PropTypes.bool,
  onSelectChange: PropTypes.func,
};
