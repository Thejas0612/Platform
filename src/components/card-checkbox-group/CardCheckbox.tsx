import { FC, MouseEventHandler } from "react";
import { Box, Checkbox, FormControlLabel, Grid, Tooltip, Typography, useTheme } from "@mui/material";
import styles from "./CardCheckbox.module.css";

export interface CardCheckboxProps {
  id: string,
  name: string
  title: string
  imageUrl: string
  disabledTooltip?: string,
  disabled?: boolean,
  onCompareChange?: () => void,
  selectedChecked?: boolean,
  comparedChecked?: boolean,
  onSelectChange?: () => void,
}

const HORIZONTAL_CENTER_TOOLTIP: Parameters<typeof Tooltip>[0]["slotProps"] = {
  popper: {
    modifiers: [
      {
        name: "offset",
        options: {
          offset: [0, -90]
        }
      }
    ]
  }
};

export const CardCheckbox: FC<CardCheckboxProps> = ({
                                                                     name,
                                                                     title,
                                                                     imageUrl,
                                                                     disabled = false,
                                                                     disabledTooltip,
                                                                     onCompareChange = () => {
                                                                     },
                                                                     selectedChecked = false,
                                                                     comparedChecked = false,
                                                                     onSelectChange = () => {
                                                                     }
                                                                   }) => {
  const theme = useTheme();
  const disabledColor = theme.palette.action.disabled;

  const handleCompareChange: Parameters<typeof Checkbox>[0]["onChange"] = (event) => {
    event.stopPropagation();
    onCompareChange();
  };

  const handleSelectChange: MouseEventHandler<HTMLDivElement> = () => {
    if (disabled) {
      return;
    }

    onSelectChange();
  };

  return <Box onClick={handleSelectChange}
              sx={{
                cursor: !disabled ? "pointer" : undefined,
                padding: ".5rem",
                border: selectedChecked ? ".125rem solid var(--ddl-color--primary-emerson-green-dark)" : ".0625rem solid var(--ddl-color--primary-grey)"
              }}>
    <Grid direction="column"
          spacing={.5}
          alignItems="center"
          container
    >
      {/* Title */}
      <Grid item>
        <Typography sx={{ color: disabled ? disabledColor : undefined }}>{title}</Typography>
      </Grid>

      {/* Image and Disabled Tooltip */}
      <Grid item>
        {disabled ?
          <Tooltip title={disabledTooltip} placement="bottom" slotProps={HORIZONTAL_CENTER_TOOLTIP}>
            <img src={imageUrl} alt={title} className={styles.cardCheckbox__image} />
          </Tooltip>
          :
          <img src={imageUrl} alt={title} className={styles.cardCheckbox__image} />}
      </Grid>

      {/* Checkbox */}
      <Grid item>
        <FormControlLabel disabled={disabled}
                          onClick={(event) => { event.stopPropagation() }}
                          sx={{ textAlign: "center" }}
                          control={<Checkbox onChange={handleCompareChange} checked={comparedChecked} />}
                          label="compare"
                          name={name} />
      </Grid>
    </Grid>
  </Box>;
};