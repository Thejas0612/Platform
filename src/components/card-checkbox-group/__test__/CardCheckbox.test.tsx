import { render, screen } from "@testing-library/react";
import { CardCheckbox, CardCheckboxProps } from "../CardCheckbox";
import userEvent from "@testing-library/user-event";

const PROPS: CardCheckboxProps = {
  id: "Test Id",
  name: "Test Name",
  title: "Test Title",
  imageUrl: "https://placehold.co/150x150"
};

describe("<CardCheckbox />", () => {
  it("when state is default, then show title, image and checkbox", () => {
    const { container } = render(<CardCheckbox {...PROPS} />);

    expect(container).toMatchSnapshot();
  });

  describe("when state is disabled", () => {
    it("then show tooltip", async () => {
      const props: CardCheckboxProps = { ...PROPS, disabled: true, disabledTooltip: "test disabled tooltip" };

      const { container } = render(<CardCheckbox {...props} />);

      const image = screen.getByRole("img", { name: /test disabled tooltip/i });
      await userEvent.hover(image);

      const tooltip = await screen.findByRole("tooltip");

      expect(tooltip).toMatchSnapshot();
      expect(container).toMatchSnapshot();
    });

    it("then prevent select", async () => {
      const onSelectChange = jest.fn();
      const props: CardCheckboxProps = {
        ...PROPS,
        disabled: true,
        disabledTooltip: "test disabled tooltip",
        onSelectChange
      };

      render(<CardCheckbox {...props} />);

      const title = screen.getByText("Test Title");
      await userEvent.click(title);

      expect(onSelectChange).not.toHaveBeenCalled();
    });
  });

});