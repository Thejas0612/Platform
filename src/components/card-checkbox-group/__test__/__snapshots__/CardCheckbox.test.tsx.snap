// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardCheckbox /> when state is default, then show title, image and checkbox 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-0.5 MuiGrid-direction-xs-column css-1r5j69-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        Test Title
      </p>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <img
        alt="Test Title"
        class="cardCheckbox__image"
        src="https://placehold.co/150x150"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-abeivz-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            name="Test Name"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          compare
        </span>
      </label>
    </div>
  </div>
</div>
`;

exports[`<CardCheckbox /> when state is disabled, then show tooltip 1`] = `
<div
  class="base-Popper-root MuiTooltip-popper MuiTooltip-popperInteractive css-yem0l2-MuiPopper-root-MuiTooltip-popper"
  data-popper-escaped=""
  data-popper-placement="bottom"
  data-popper-reference-hidden=""
  id="mui-1"
  role="tooltip"
  style="position: absolute; top: 0px; left: 0px; margin: 0px; transform: translate(0px, -90px);"
>
  <div
    class="MuiTooltip-tooltip MuiTooltip-tooltipPlacementBottom css-ja5taz-MuiTooltip-tooltip"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 133ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    test disabled tooltip
  </div>
</div>
`;

exports[`<CardCheckbox /> when state is disabled, then show tooltip 2`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-0.5 MuiGrid-direction-xs-column css-1r5j69-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1qhoxxv-MuiTypography-root"
      >
        Test Title
      </p>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <img
        alt="Test Title"
        aria-label="test disabled tooltip"
        class="cardCheckbox__image"
        data-mui-internal-clone-element="true"
        src="https://placehold.co/150x150"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <label
        class="MuiFormControlLabel-root Mui-disabled MuiFormControlLabel-labelPlacementEnd css-abeivz-MuiFormControlLabel-root"
      >
        <span
          aria-disabled="true"
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium Mui-disabled PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium Mui-disabled MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          tabindex="-1"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            disabled=""
            name="Test Name"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label Mui-disabled css-ahj2mt-MuiTypography-root"
        >
          compare
        </span>
      </label>
    </div>
  </div>
</div>
`;
